library/src/utility/device_memory.cpp

```diff
@@ -14,16 +15,38 @@ void* DeviceMem::GetDeviceBuffer() const { return mpDeviceBuf; }
 
 std::size_t DeviceMem::GetBufferSize() const { return mMemSize; }
 
-void DeviceMem::ToDevice(const void* p) const
+void DeviceMem::ToDeviceImpl(const void* p) const
 {
     hip_check_error(hipMemcpy(mpDeviceBuf, const_cast<void*>(p), mMemSize, hipMemcpyHostToDevice));
 }
 
-void DeviceMem::FromDevice(void* p) const
+void DeviceMem::FromDeviceImpl(void* p) const
 {
     hip_check_error(hipMemcpy(p, mpDeviceBuf, mMemSize, hipMemcpyDeviceToHost));
 }
 
+#ifdef CK_EXPERIMENTAL_BIT_INT_EXTENSION_INT4
+void DeviceMem::ToDevice(const ck::int4_t* host) const
+{
+    const std::size_t count = GetBufferSize();
+    const auto buffer       = std::make_unique<int8_t[]>(count);
```

Comment:
I don't think converting int4 to int8 behind ToDevice() is appropriate. This is a workaround after all.

I can imagine this is for streamlining the example code. But I think it's better to have int4->int8 conversion explicit in the example code, so we can better avoid someone getting tripped by the magic ToDevice(const int4_t*) down the line.

Reply:
Related comment:

> The bit-pattern of `ck::int4_t` and `int8_t` is not same, thus we cannot simply use `ck::int4_t` buffer as `int8_t` before transfer data between host & device. Shall add code to converting values.

Can we use some intermediate Tensor<int8_t> object for converting user int4 data to kernel-supported int8 data, then copy the int8 data to DeviceMem object without going through the magic ToDevice(const int4_t*)?

Reply:
I agree with you. Moving the conversion logics into `DeviceMem` is error-prone if we change how we support `ck::int4_t` in the future.